### PR TITLE
Add custom map container for mapping from MOOSE variable numbers to any type

### LINKED ISSUE #4624: add MooseVariableMap
I need a map object that allows quick value lookup for a given Moose Variable number. Naive as I am I previously thought I could just make an ```std::vector``` (as Moose Variable numbers are unsigned ints), but I hadn't thought about AuxVariables, which get assigned numbers of MAX_UINT-n.

Using an ```std::map``` has been a solution, but the performance overhead is prohibitive, as this lookup is performed for every ```computeQpOffDiagJacobian``` call. The map lookup takes way more time than the rest of the Jacobian evaluation (profiled).

To fix this I went back to a direct ```std::vector``` lookup, but I map NL vars to even numbers (starting at 0) and AuxVars to odd numbers (starting at 1). This way The lookup table is at most 50% empty and the lookup times are reduced by an order of magnitude.

I'd like to make this "map" type available as an Interface that can be inherited from. This map would then also be used in the existing ```JvarMapInterface``` instead of ```std::map``` (or the LIBMESH_FASTED_...MAP)

### PR DESCRIPTION
**Pull request by dschwen**
Custom container to hold maps form Moose Variable numbers to arbitrary other objects. This container exploits the numbering scheme that MOOSE and libMesh use for the non-linear and auxiliary variables and avoids tree or hash based mapping in favor of a faster vector based lookup.

Closes #4624.

### CODE REVIEW

File: framework/include/utils/MooseVariableMap.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+private:
+  /**
+   * This function remaps the the Non-linear (and Aux-Variable) numbers
+   * to even (and odd) numbers starting at zero (and one).
+   * This allows us to use a fairly dense vector to perform the direct lookup.
+   */
+  unsigned int libMeshVarNumberRemap(unsigned int var) const
+  {
+    const int b = static_cast<int>(var);
+    const int i = b >= 0 ? b<<1 : (-b<<1)-1;
```
Comment (permcody) [inline framework/include/utils/MooseVariableMap.h]: Oooh, fancy!

**Comment by permcody** [comment]
Based on the timings done by @jwpeterson. I'd like more feedback from the rest of the @idaholab/moose-developers. 

This is a fairly clean class so it's probably fine, but it sucks that it's only necessary on our Mac workstations when we us clang. Those that care could just not use clang... See #4624 for the full discussion and timings.

**Comment by dschwen** [comment]
That last statement is inaccurate. The std::map vs. std::unorderd_map issue has probably side tracked you. The important comparison is _this class_ vs. _the faster of map and unordered_map_ and this one still wins easily.

**Comment by roystgnr** [comment]
What the hell, Apple clang?  Should we be scouring through libMesh for low-N unordered_map cases so we can revert them to map on lousy compiler versions?

The numbering-convention-aware remapping here is much more clever than anything you could tease out of a standard container, so I think "only necessary with Mac/clang" is probably an overstatement.

**Comment by jwpeterson** [comment]
> What the hell, Apple clang? Should we be scouring through libMesh for low-N unordered_map cases so we can revert them to map on lousy compiler versions?

This is a good idea.  If there is enough information available at configure time I'd be in favor of making `LIBMESH_BEST_UNORDERED_MAP` a `std::map` on Linux/clang architectures regardless of N.

I also think we should probably merge this PR regardless.  While I do appreciate @permcody's desire to avoid adding a special kind of hand-coded map for every possible situation, this one seems generally useful enough to be worth it.

**Comment by permcody** [comment]
I asked for a vote. I'm not completely against this. If there are a few
that want this container, then merge it.

**Comment by andrsd** [comment]
@dschwen Do you have a comparison that shows your class is faster that any std-container? Maybe I missed it... Also, is your class thread-safe? Does it have to be?

**Comment by dschwen** [comment]
Yeah, in the issue. The thread safety is addressed there as well.

**Comment by friedmud** [comment]
I'm currently voting against this.  Either the algorithm that needs this should be changed... or maybe this should be more localized to the Kernels that need it.  @YaqiWang hasn't needed something like this and he has very similar problems to you guys.  Have you talked to @YaqiWang about how he approaches this problem?

@dschwen Where can I see some code that is currently using std::map?  I'd like to see what your algorithm looks like.

**Comment by dschwen** [comment]
@friedmud this is currently used in in a few places where we get an ```unsigned int j_var``` variable number passed in and need to look up a certain property from that. We have this in Kernels (```SpliCHParsed```, ```ACParsed```) in the ```computeQpOffDiagJacobian``` calls and in certain derivative materials (```DerivativeTwoPhasMaterial```, ```ElasticEnergyMaterial```, and the multiphase stuff I'm working on) in the ```computeDF``` etc. methods (for which we modelled the interface after the off diag Jacobian method).

The common scenario is always fetching a table of all derivatives that are needed (while looping over the coupled variables) in the CTOR (here we index into the vector of coupled variables), and then pulling from that table for a given variable number (where the variable number is _not_ an index into teh vector of coupled variables).

Note, that the ```argIndex()``` stuff from the Materials is currently horribly broken for AuxVariables, but this is fixed in my multiphase_4545 branch.

I'd be happy to get suggestions on how to approach this differently.

**Comment by friedmud** [comment]
We never ask for the derivative with respect to Auxiliary variables... so why all of the machinery to deal with them?

**Comment by dschwen** [comment]
Damn it. That is actually a good point. Why would ```computeQpOffDiagJacobian``` be called for an AuxVar? Similarly I could have ```DerivativeBaseMaterial``` skip AuxVars when calling the ```computeDF``` methods!

**Comment by permcody** [comment]
I guess we don't need this fancy container anymore? Still learned a lot about broken compilers though in the process.

**Comment by YaqiWang** [comment]
hehe, I just wanted to jump in and saw your two's conversation. The derivative stuff should be done at a more base level in my opinion. The quick and dirty solution can exist if not a lot of people ask for this functionality. Since moose use jvar, i think we all should just use it for indexing. Additional mapping seems to me not quite necessary.

**Comment by dschwen** [comment]
Yes, @YaqiWang , you are probably right. The additional mapping is not really necessary if we can rely on contiguous variable numbers and if we can ignore aux (and their O(max_int) numbers).

**Pull request closed without merging**